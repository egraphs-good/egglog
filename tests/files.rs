use egg_smol::*;

fn test_program(program: &str, message: &str) {
    let mut egraph = EGraph::default();
    match egraph.parse_and_run_program(program) {
        Ok(msgs) => {
            for msg in msgs {
                log::info!("  {}", msg);
            }
        }
        Err(err) => panic!("{}: {err}", message),
    }
}

fn run(path: &str) {
    let _ = env_logger::builder().is_test(true).try_init();
    let program = std::fs::read_to_string(path).unwrap();
    test_program(&program, "Top level error");

    let egraph = EGraph::default();
    let program_str = egraph
        .parse_program(&program)
        .unwrap()
        .into_iter()
        .map(|x| x.to_string())
        .collect::<Vec<String>>()
        .join("\n");
    test_program(
        &program_str,
        &format!(
            "Program:\n{}\n ERROR after parse, to_string, and parse again.",
            program_str
        ),
    );
}

// include the tests generated from the build script
include!(concat!(std::env!("OUT_DIR"), "/files.rs"));
