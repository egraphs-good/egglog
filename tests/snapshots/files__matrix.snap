---
source: tests/files.rs
expression: serialized
---
nodes:
  Kron-18295084334957702260:
    op: Kron
    children:
      - MMul-2045059113437825509
      - MMul-9274558654761462362
    eclass: "10"
    cost: 1
  Kron-7916840120001827962:
    op: Kron
    children:
      - Id-3429551472952562336
      - B-0
    eclass: "39"
    cost: 1
  Kron-14865532862005139924:
    op: Kron
    children:
      - MMul-2045059113437825509
      - Id-9301332479516564789
    eclass: "41"
    cost: 1
  Kron-11346391592954390298:
    op: Kron
    children:
      - MMul-9683589484919301784
      - MMul-17323788914821161860
    eclass: "60"
    cost: 1
  Kron-8959473804624910437:
    op: Kron
    children:
      - Id-16108461796980496807
      - C-0
    eclass: "69"
    cost: 1
  m-0:
    op: m
    children: []
    eclass: "33"
    cost: 9223372036854776000
  B-0:
    op: B
    children: []
    eclass: "26"
    cost: 9223372036854776000
  C-0:
    op: C
    children: []
    eclass: "5"
    cost: 9223372036854776000
  Times-16173041864095337072:
    op: Times
    children:
      - NamedDim-5485299131631678697
      - NamedDim-5485299131631678697
    eclass: "78"
    cost: 1
  Times-4775599672430715124:
    op: Times
    children:
      - n-0
      - m-0
    eclass: "77"
    cost: 1
  Times-10616345707667886544:
    op: Times
    children:
      - NamedDim-13708613989203058623
      - NamedDim-9813572419544689236
    eclass: "77"
    cost: 1
  MMul-17860948738465144212:
    op: MMul
    children:
      - Kron-7916840120001827962
      - Kron-14865532862005139924
    eclass: "10"
    cost: 1
  MMul-2045059113437825509:
    op: MMul
    children:
      - Id-3429551472952562336
      - MMul-2045059113437825509
    eclass: "25"
    cost: 1
  MMul-9274558654761462362:
    op: MMul
    children:
      - MMul-9274558654761462362
      - Id-9301332479516564789
    eclass: "26"
    cost: 1
  MMul-17323788914821161860:
    op: MMul
    children:
      - Id-9301332479516564789
      - MMul-17323788914821161860
    eclass: "58"
    cost: 1
  MMul-9683589484919301784:
    op: MMul
    children:
      - Id-3429551472952562336
      - MMul-9683589484919301784
    eclass: "48"
    cost: 1
  MMul-1960381455249418021:
    op: MMul
    children:
      - Kron-7916840120001827962
      - Kron-11346391592954390298
    eclass: "39"
    cost: 1
  MMul-9103629317315635186:
    op: MMul
    children:
      - MMul-3283821939192825387
      - Kron-14865532862005139924
    eclass: "41"
    cost: 1
  MMul-3283821939192825387:
    op: MMul
    children:
      - MMul-3283821939192825387
      - Kron-11346391592954390298
    eclass: "60"
    cost: 1
  MMul-1445829347140181091:
    op: MMul
    children:
      - Kron-8959473804624910437
      - MMul-9103629317315635186
    eclass: "70"
    cost: 1
  MMul-7265636725262990890:
    op: MMul
    children:
      - Kron-8959473804624910437
      - MMul-3283821939192825387
    eclass: "73"
    cost: 1
  ex1-0:
    op: ex1
    children: []
    eclass: "10"
    cost: 9223372036854776000
  String-9813572419544689236:
    op: "\"n\""
    children: []
    eclass: String-9813572419544689236
    cost: 0
  NamedDim-9813572419544689236:
    op: NamedDim
    children:
      - String-9813572419544689236
    eclass: "32"
    cost: 1
  String-13708613989203058623:
    op: "\"m\""
    children: []
    eclass: String-13708613989203058623
    cost: 0
  NamedDim-13708613989203058623:
    op: NamedDim
    children:
      - String-13708613989203058623
    eclass: "33"
    cost: 1
  String-5485299131631678697:
    op: "\"p\""
    children: []
    eclass: String-5485299131631678697
    cost: 0
  NamedDim-5485299131631678697:
    op: NamedDim
    children:
      - String-5485299131631678697
    eclass: "75"
    cost: 1
  Id-3429551472952562336:
    op: Id
    children:
      - NamedDim-9813572419544689236
    eclass: "48"
    cost: 1
  Id-9301332479516564789:
    op: Id
    children:
      - NamedDim-13708613989203058623
    eclass: "58"
    cost: 1
  Id-16108461796980496807:
    op: Id
    children:
      - NamedDim-5485299131631678697
    eclass: "68"
    cost: 1
  n-0:
    op: n
    children: []
    eclass: "32"
    cost: 9223372036854776000
root_eclasses: []
class_data:
  "10":
    type: MExpr
  "25":
    type: MExpr
  "26":
    type: MExpr
  "39":
    type: MExpr
  "48":
    type: MExpr
  "41":
    type: MExpr
  "58":
    type: MExpr
  "60":
    type: MExpr
  "69":
    type: MExpr
  "68":
    type: MExpr
  "5":
    type: MExpr
  "33":
    type: Dim
  "78":
    type: Dim
  "75":
    type: Dim
  "77":
    type: Dim
  "32":
    type: Dim
  "70":
    type: MExpr
  "73":
    type: MExpr
  String-9813572419544689236:
    type: String
  String-13708613989203058623:
    type: String
  String-5485299131631678697:
    type: String

